lxd/storage/drivers/driver_zfs_utils.go

```diff
@@ -123,33 +123,78 @@ func (d *zfs) deleteDatasetRecursive(dataset string) error {
 
 	// Check if the origin can now be deleted.
 	if origin != "" && origin != "-" {
-		if strings.HasPrefix(origin, d.config["zfs.pool_name"]+"/deleted") {
-			// Strip the snapshot name when dealing with a deleted volume.
-			dataset, _, _ = strings.Cut(origin, "@")
-		} else if strings.Contains(origin, "@deleted-") || strings.Contains(origin, "@copy-") {
-			// Handle deleted snapshots.
-			dataset = origin
-		} else {
-			// Origin is still active.
-			dataset = ""
+		// It will be set non-empty if origin (or parent of) should be deleted.
+		var originToDelete string
+
+		// Determine if the origin is a temporary snapshot created.
+		IsOriginTemporarySnapshot := strings.Contains(origin, "@deleted-") || strings.Contains(origin, "@copy-")
+
+		// Determine deletino strategy based on origin type and location:
+		// - Temporary snapshots (deleted- or copy-) should be deleted themselves.
+		// - Regular snapshots in deleted/ dataset shuould cause parent dataset deletion (when dataset is in deleted/, it's marked for cleanup, so we clean up the parent dataset when its last snapshot is removed).
+
+		if IsOriginTemporarySnapshot {
+			// Origin is a temporary snapshot fot this dataset, delete it.
+			originToDelete = origin
+		} else if strings.HasPrefix(origin, d.config["zfs.pool_name"]+"/deleted/") {
+			// This is a regular snapshot for dataset in deleted/ dataset.
+			// Delete the parent dataset (strip the @part) because when a dataset is in deleted/, it's marked for cleanup, and we should remove the entire dataset when its last snapshot is gone.
+			originToDelete, _, _ = strings.Cut(origin, "@")
 		}
 
-		if dataset != "" {
-			// Get all clones.
-			clones, err := d.getClones(dataset)
-			if err != nil {
-				return err
-			}
+		// If no origin needs to be deleted we're done.
+		if originToDelete == "" {
+			return nil
+		}
+
+		// Check if the origin (or parent dataset) has any clones.
+		clones, err := d.getClones(originToDelete)
+		if err != nil {
+			return err
+		}
+
+		// Only proceed with deletion if there are no clones.
+		if len(clones) == 0 {
```

Comment:
we can do happy path here to avoid indention, e.g. `if len(clones) >0 { return nil }`